EVA-1634 — Handle multiple alignments

mistyskye opened the pull request:
- Added flanking sequence length support in reverse_strand.py as a parameter: -f
- Fixed a few errors
- Updated README

mistyskye committed:
Initial commit with flanking sequence length as a paramter

- Added flanking sequence length as a parameter: -f
- Fixed a few errors
- Updated README

mistyskye committed:
Added flanking sequence length as a parameter, various fixes

- Flanking sequence length can now be chosen with `-f`
- The pipeline now filters out reads if they are shorter than the required length
- Fixed a problem where unmapped reads were not being filtered out

tskir commented on reverse_strand.py:
```diff
@@ -26,14 +28,14 @@
     name = read.query_name
     info = name.split("|")
     nucl = info[3]
+    if read.is_unmapped: # Can be decoded with bitwise flag with & 4 (4 means unmapped)
```
Minor style comment: per PEP 8 style guide, comment should be separated by two spaces from the code, e. g. :

```python
    if read.is_unmapped:  # Can be decoded with bitwise flag with & 4 (4 means unmapped)
```

(Here and throughout the code)

tskir commented on reverse_strand.py:
```diff
@@ -6,18 +6,20 @@
 bamfile = pysam.AlignmentFile(args.bam, 'rb')
 outfile = open(args.outfile, 'w')
 old_ref_alleles = open(args.old_ref_alleles, 'w')
+flanklength=args.flankingseqlength
```
Spaces around the assignment operator

tskir commented on remapping_commands.sh:
```diff
@@ -118,7 +121,7 @@ echo '------------------------------3) Data extraction--------------------------
+../reverse_strand.py -i reads_aligned.sorted.bam -p old_ref_alleles.txt -o variants_remapped.vcf -f "$flankingseq"
 
 # Add interval for variant position in bed format (required by getfasta)
 # Reprints all the columns, adding an extra column before the pos column as the pos-1, as bedtools getfasta requires a 
```
AS/XS filtering is not implemented in the current version of the code in this PR (just adding this discussion in order to not forget it before closing the ticket)

mistyskye committed:
Added filtering using AS and XS, and based on local region alignment

- The pipeline now filters based on a minimum AS (alignment score), and a minimum AS difference between the best alignment and second-best alignment
- It also checks the region around the variant position, and only keeps variants that have a perfect alignment (matches or mismatches) in that region (default length to be determined)
- Fixed a couple of style suggestions

mistyskye committed:
Changed penalty for N alignments, added uniq filter, changed AS cut-off calculation, added counts

- Ns no longer have a penalty (penalty=0), so that reads with a few Ns don't have a low AS (--np 0)
- Added a uniq filter just after the SNP filter to be able to calculate comparison stats on the output file
- AS cut-off is now calculated based on flanking seq length, aka read length
- reverse_strand.py now calculates and prints filter stats

tskir commented on remapping_commands.sh:
```diff
@@ -26,47 +30,57 @@ cut -f1,2 ../"$oldgenome".fai > "$oldgenome".chrom.sizes
+# Get unique variants based on rsIDs (column 3)
+grep '^#' snps_only.vcf > snps_only.uniq.vcf
+grep -v '^#' snps_only.vcf | sort -u -k3,3  >> snps_only.uniq.vcf
+
```
This step should be removed (or at least commented out, for now), because _in general_ we don't want to change a user's input VCF more than necessary. The variant ID being present multiple lines in a VCF is a problem **for us** because it makes it harder to calculate summary statistics, but theoretically if a user has VCF with repeating identifiers, we should leave them as is. Also consider the possibility that a user's VCF might lack identifiers altogether; this will remove _all but one record_

tskir commented on remapping_commands.sh:
```diff
@@ -26,47 +30,57 @@ cut -f1,2 ../"$oldgenome".fai > "$oldgenome".chrom.sizes
-gzip -d snps_only.vcf.gz
-vcf2bed < snps_only.vcf > variants.bed
+gzip -d snps_only.uniq.vcf.gz
+vcf2bed < snps_only.uniq.vcf > variants.bed
```
After removing/commenting out lines about uniq filtering, these should be changed back as well

tskir commented on remapping_commands.sh:
```diff
@@ -81,7 +95,7 @@ paste <(grep '^>' variants_read_filtered.fa) old_ref_bases.txt variant_bases.txt
 # [seq]
 sed 's/\t/|/g; s/ /|/g; s/\(.*\)|/\1 /' temp.txt | tr ' ' '\n' > variant_reads.out.fa
-
+#total=$(grep "^[^>]" variant_reads.out.fa | wc -l)
```
Seeing as the line is commented out, is it actually used anywhere?

tskir commented on remapping_commands.sh:
```diff
@@ -94,14 +108,16 @@ echo "#### Mapping results: ####"
 
 # Bowtie2 arguments:
 # -k 1 means that bowtie2 will only report the best alignment for each variant (default behaviour)
+# -k 2 means that bowtie2 will report the best alignment and the second best
```
Do we actually need the second best alignment itself? I was under the impression that even for `-k 1` mode Bowtie reports an XS score, which is all we need

mistyskye replied:
[From the docs](http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#k-mode-search-for-one-or-more-alignments-report-each) it seems that `-k 1` is the default mode whether it's specified or not, [which only reports 1 alignment](http://bowtie-bio.sourceforge.net/manual.shtml#example-4-default--k-1). So I think although it may report an XS, it's always the same as the AS because it doesn't actually look for a 2nd alignment.

tskir replied:
OK, let's just leave it at `-k 2`, it shouldn't do any harm anyway

tskir commented on reverse_strand.py:
```diff
@@ -6,18 +6,31 @@
 
 description = "Reverses the allele for variants that got mapped onto the reverse strand of the new genome, and prints"
-            "everything to a new file\n"
+"everything to a new file\n"
```
This second line would not be included in the description because there are neither backslash not parentheses present; `"everything to a new file\n"` will evaluate as a separate string literal which is not used anywhere. You can use this syntax to make automated literal concatenation work:

```python
description = ("Reverses the allele for variants that got mapped onto the reverse strand of the new genome, and "
               "prints everything to a new file\n")
```

tskir commented on reverse_strand.py:
```diff
@@ -26,14 +39,72 @@
+    try:
+        XS = read.get_tag("XS")  # Some variants don't have secondary alignments, which throws an error
+    except KeyError:
+        XS = -int(flanklength) # Set an arbitrary low value for the "artificial" secondary alignment 
```
PEP8: double space before a `#`

tskir commented on reverse_strand.py:
```diff
@@ -26,14 +39,72 @@
+
+    # Calculating correct position and filtering based on alignement around the variant position
+    start = read.pos
+    readVarPos = start + int(flanklength) + 1
```
PEP8: camelCase is not used in Python for naming variables; it should be `ready_var_pos` instead (here and throughout the Python modules)

https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

tskir commented on reverse_strand.py:
```diff
@@ -26,14 +39,72 @@
+        # Stop incrementing varpos once we've reached readVarPos:
+        if start + operator < readVarPos:
+            # Mismatch/matches and deletions increase the position counter:
+            if cigList[operator] == 0 or cigList[operator] == 2: 
```
Instead of 0 and 2, you can use constants `BAM_CMATCH` and `BAM_CDEL` instead

tskir commented on reverse_strand.py:
```diff
@@ -26,14 +39,72 @@
+        # If we are in the local region around the variant, but not at the position of the variant:
+        if (start + operator >= (readVarPos - localRegionSize)) \
+        and (start + operator <= (readVarPos + localRegionSize)) \
+        and (start + operator != (readVarPos)):
```
As per PEP 8, this should be formatted:

```python
        if (start + operator >= (readVarPos - localRegionSize))
            and (start + operator <= (readVarPos + localRegionSize))
            and (start + operator != (readVarPos)):
```

(Also note that the backslash it not required if you're using parentheses; they create an implied line continuation)

https://www.python.org/dev/peps/pep-0008/#indentation

tskir commented on reverse_strand.py:
```diff
@@ -26,14 +39,72 @@
+                perfCounter += 1  # Increase the counter for perfect local region
+    if (AS <= int(scorecutoff)):
+        primaryPoorCount += 1
+    elif (AS - XS < 20):
```
This difference (20) should not be hardcoded and must instead be either a constant or a user-defined parameter. Also, given that AS and XS are both linearly (ish) dependent on flanking sequence length, this should probably be also specified as a fraction of that length

tskir commented on reverse_strand.py:
```diff
@@ -26,14 +39,72 @@
+print(remappedCount)
+print("% of variants rejected for:")
+print("Unmapped:", round(((unmappedCount/totalCount)*100),3), "%")
+print("Primary alignment too poor:", round(((primaryPoorCount/totalCount)*100), 3), "%")
```
You don't have to do rounding and formatting manually; there is a very handy string formatting capability in Python (see https://www.learnpython.org/en/String_Formatting for example). In this scenario, you can use something like this:

```python
print("Primary alignment too poor: {:.1%}".format(primaryPoorCount/totalCount))
```

mistyskye committed:
Changed a few things, updated README, fixed style issues

- The pipeline doesn't require an accession ID in the command line (it just uses the "new" genome filename)
- Updated README to describe new parameters
- Fixed formatting and style problems
- Commented the uniq commands
